src/viam/sdk/robot/client.cpp

```diff
@@ -228,9 +322,9 @@ RobotClient::RobotClient(std::shared_ptr<ViamChannel> channel)
       should_close_channel_(false),
       stub_(RobotService::NewStub(channel_)) {}
 
-std::vector<ResourceName>* RobotClient::resource_names() {
+std::vector<Name>* RobotClient::resource_names() {
     const std::lock_guard<std::mutex> lock(lock_);
-    std::vector<ResourceName>* resources = &resource_names_;
+    std::vector<Name>* resources = &resource_names_;
```

Comment:
Oh, this is worse than I thought. The lock is completely defeated here because the returned pointer is to internal state. This needs to return by value.

Reply:
Not entirely certain I follow why this needs to return by value, does changing return type to `const&` not solve this? If so, then returning a copy seems totally fine.

Reply:
No, returning by `const&` isn't sufficient, because you could still be doing an (unlocked) read of the state via the reference at the same time a locked mutation was in progress, since the thread holding the reference doesn't hold the lock.